src/mito2/src/gc.rs

```diff
@@ -0,0 +1,420 @@
+                        removed_files.extend(action.files_to_remove.clone());
+                        manifest_builder.apply_edit(manifest_version, action);
+                    }
+                    RegionMetaAction::Remove(_) => {}
+                    RegionMetaAction::Truncate(action) => {
+                        match &action.kind {
+                            crate::manifest::action::TruncateKind::All { .. } => {
+                                removed_files.extend(manifest_builder.files().values().cloned());
+                            }
+                            crate::manifest::action::TruncateKind::Partial { files_to_remove } => {
+                                removed_files.extend(files_to_remove.clone());
+                            }
+                        }
+                        manifest_builder.apply_truncate(manifest_version, action);
+                    }
+                }
+            }
+            ret.entry(ts)
+                .or_insert_with(HashSet::new)
+                .extend(removed_files);
+        }
+
+        Ok(ret)
+    }
+
+    /// Concurrently list unused files in the region dir
+    /// because there may be a lot of files in the region dir
+    /// and listing them may take a long time.
+    pub async fn list_unused_files(
+        &self,
+        region_id: RegionId,
+        in_used: HashSet<FileId>,
+        recently_removed_files: BTreeMap<Option<Timestamp>, HashSet<FileMeta>>,
+        concurrency: usize,
+    ) -> Result<Vec<Entry>> {
+        let may_linger_until =
+            chrono::Utc::now() - chrono::Duration::from_std(self.opt.lingering_time).unwrap();
+
+        let unknown_file_may_linger_until = chrono::Utc::now()
+            - chrono::Duration::from_std(self.opt.unknown_file_lingering_time).unwrap();
+
+        // files that may linger, which means they are not in use but may still be kept for a while
+        let may_linger_filenames = recently_removed_files
+            .iter()
+            .filter_map(|(ts, files)| {
+                if let Some(ts) = ts {
+                    if *ts < Timestamp::new_millisecond(may_linger_until.timestamp_millis()) {
+                        // if the expel time is before the may linger time, we can delete it
+                        return None;
+                    }
+                }
+                Some(files)
+            })
+            .flatten()
+            .flat_map(|meta| {
+                [
+                    location::sst_file_path(
+                        self.access_layer.table_dir(),
+                        meta.file_id(),
+                        self.access_layer.path_type(),
+                    ),
+                    location::index_file_path(
+                        self.access_layer.table_dir(),
+                        meta.file_id(),
+                        self.access_layer.path_type(),
+                    ),
+                ]
+            })
+            .collect::<HashSet<_>>();
```

Comment:
Will it be wasteful to keep all path strings in the set? Can we only maintain a `HashSet<FileId>`? We list and parse the files to check their file id.